### PR TITLE
Push manager. Fixes issue #2

### PR DESCRIPTION
**Pull request by efullea**
Changes:
-Move pushRegistrationManager to ServiceWorkerRegistration (Fixes issue #2)
-PushManager --> PushRegistrationManager
-Adding PushRegistrationManager.hasPermission

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -802,6 +790,36 @@
       </li>
    </ol>
      
+  <p> The <dfn><code>hasPermission</code></dfn> method when invoked MUST run the
+  following steps:
+    <ol>
+      <li>Let <var>promise</var> be a new <a><code>Promise</code></a> object and
```
Comment (beverloo) [inline index.html]: @mvano is changing the steps to not explicitly mention |resolver| in https://github.com/w3c/push-api/pull/23. He should probably rebase his patch on top of this.

File: index.html [hunk 2]
```diff
@@ -1030,7 +1048,31 @@
...
+              permission to use Push API.
+          </dd>
+
+          <dt>denied</dt>
+          <dd>
+              The <a class="internalDFN" href="#dfn-webapp">webapp</a> has been
+              denied permission to use Push API.
+          </dd>
+
+          <dt>needToAsk</dt>
```
Comment (beverloo) [inline index.html]: The Web Notification API uses "default" for this, Mounir proposed a Permissions API to DAP today that uses "prompt", you're introducing "needToAsk" :-).

Could we settle on a single keyword? While I'm a fan of "prompt", "default" is already out there.

@mounirlamouri

File: index.html [hunk 3]
```diff
@@ -802,6 +790,36 @@
...
+      algorithm with no arguments and terminate these steps.
+      </li>
+      <li>When the request has been completed invoke <em>resolver</em>'s <a class="internalDFN"
+          href="#dfn-fulfill-algorithm">fulfill algorithm</a> algorithm with
+          the <a><code>PushPermissionStatus</code></a> as the <code>value</code> argument.
+        </ol>
+      </li>
+   </ol>
+
+  <p> Permission to use the Push Service can be persistent, that is, it does not
```
Comment (beverloo) [inline index.html]: nit: s/Push Service/push service/. We don't capitalize it elsewhere.

File: index.html [hunk 4]
```diff
@@ -656,6 +637,13 @@
 		notified about the result of the operation. 
     </dd>
     
+    <dt> Promise&lt;PushPermissionStatus&gt; hasPermission ()</dt>
+    <dd>This method allows a <a class="internalDFN"
+    href="#dfn-webapp">webapp</a> to check whether user has granted permission
+    to use push service. It returns a <a><code>Promise</code></a> that will
+    allow the caller to be notified about the result of the operation.
```
Comment (beverloo) [inline index.html]: ".. that will notify the caller about the result of the operation"? The proposed text makes it sound like the caller MAY want to do this, while the method is a no-op unless the result gets used.

**Comment by beverloo** [comment]
Firstly, thank you for waiting!! :-)

While I don't remember my exact concern, one thing I do care about is availability of the PushRegistrationManager interface from within Service Workers. A possible scenario would be unregistering the active Push registration when the server sends a message that it's no longer necessary.

The ServiceWorkerRegistration interface is explicitly only exposed on Window, which means that it will no longer be able to use Push functionality (aside from removing messages) from within a worker. There is an open issue for this in the Service Workers repository:

https://github.com/slightlyoff/ServiceWorker/issues/421

It is noteworthy that @nikhilm points out in comment 12 that he would like this done after SW v1, and that Gecko won't expose push to SWs in the first place. We do want to expose this interface within workers.

File: index.html (hunk 2 above)
Comment (efullea) [inline index.html]: 'prompt' looks more self-explanatory to me than 'default', so I would go for it.

Comment (mounirlamouri) [inline index.html]: I think _default_ was a safe word for _prompt_ in Notifications.

File: index.html (hunk 4 above)
Comment (mounirlamouri) [inline index.html]: +1. This prose is actually missing requirement wording. It is better to explicitly say what what the UA should do and whether it is MUST or SHOULD (I assume that those are the options here).

Comment (efullea) [inline index.html]: Other methods seem to have the same wording. Let's deal with it separately.

**Pull request merged**